plasmapy/utils/decorators/helpers.py

```diff
@@ -1,12 +1,105 @@
 """
 Miscellaneous decorators for various package uses.
 """
-__all__ = ["preserve_signature"]
+__all__ = ["modify_docstring", "preserve_signature"]
 
 import functools
 import inspect
 
 
+def modify_docstring(func=None, prepend: str = None, append: str = None):
+    """
+    A decorator which programmatically prepends and/or appends the docstring
+    of the decorated method/function.  The unmodified/original docstring is
+    saved as the ``__original_doc__`` attribute.
+
+    Parameters
+    ----------
+    func: callable
+        The method/function to be decorated.
+
+    prepend: `str`
+        The string to be prepended to the ``func``'s docstring.
+
+    append: `str`
+        The string to be appended to the ``func``'s docstring.
+
+    Returns
+    -------
+    callable
+        Wrapped version of the function.
+
+    Examples
+    --------
+
+        >>> @modify_docstring(prepend='''Hello''', append='''World''')
+        ... def foo():
+        ...     '''Beautiful'''
+        ...     pass
+        >>> foo.__original_doc__
+        'Beautiful'
+        >>> foo.__doc__
+        'Hello\\n\\nBeautiful\\n\\nWorld'
+
+    """
+
+    def decorator(f):
+        sig = inspect.signature(f)
+
+        @preserve_signature
+        @functools.wraps(f)
+        def wrapper(*args, **kwargs):
+            # combine args and kwargs into dictionary
+            bound_args = sig.bind(*args, **kwargs)
+            bound_args.apply_defaults()
+
+            return f(*bound_args.args, **bound_args.kwargs)
+
+        if prepend is None and append is None:
+            raise TypeError(
+                "Decorator @modify_docstring() missing argument 'prepend' and/or"
+                " 'append', at least one argument is required."
+            )
+
+        # save the original docstring
+        setattr(wrapper, "__original_doc__", wrapper.__doc__)
+        doclines = inspect.cleandoc(wrapper.__doc__).splitlines()
+
+        # prepend docstring lines
+        if isinstance(prepend, str):
+            prependlines = inspect.cleandoc(prepend).splitlines()
+            prependlines.append("")
+        elif prepend is None:
+            prependlines = []
+        else:
+            raise TypeError(
+                f"Expected type str for argument 'prepend', got {type(prepend)}."
+            )
+
+        # append docstring lines
+        if isinstance(append, str):
+            appendlines = inspect.cleandoc(append).splitlines()
+            appendlines = [""] + appendlines
+        elif append is None:
+            appendlines = []
+        else:
+            raise TypeError(
+                f"Expected type str for argument 'append', got {type(append)}."
+            )
+
+        # define new docstring
+        wrapper.__doc__ = "\n".join(prependlines + doclines + appendlines)
```

Comment:
This decorator has the potential to get significantly more complex, so I would suggest break up the prepending and appending (prependation and appendation?) into different private functions that can be called inside the decorator. 

It could be something vaguely along the lines of:
```Python
doclines = _save_original_docstring(...)
prependlines = _prepend_docstring_lines(prepend)
appendlines = _append_docstring_lines(append)
wrapper.__doc__ = _define_new_docstring(prependlines, appendlines, doclines)
```
where the comments describing the section of code become the function names (though there's probably a better way to break it up than the above code block).  I think of this pattern of refactoring when there are sections of code separated by comments.  I think I originally read it in _Clean Code_ by R. Martin.  We could also use the strategy of dropping one level of abstraction per function.

I'm also mentioning this because if I had followed this advice with the original implementation of `plasmapy.utils.pytest_helpers.run_test`, I wouldn't have needed to go through with the test helper refactoring that I'm doing in #919 through #929...oof.

Reply:
Yeah, I do have several ideas to tackle more complex scenarios.  I'll probably migrate this to a class style decorator, much like we do for `validate_quantities` and `ValidateQuantites`.  Once there, I'll have functionality for injecting into specific `numpydoc` sections.

However,  I want to leave that for the future.  I want to get this through so I can get PR #908 merged and move on with the diagnostics.  That is much more important to me at this point.

Reply:
Sure!  I'm happy to save that for the future, in particular if we create an issue or two on GitHub so that we can keep track of it for the future.